cwl/src/main/scala/wdl4s/cwl/CwlFile.scala

```diff
@@ -0,0 +1,262 @@
+import wdl4s.wom.executable.Executable
+import wdl4s.wom.expression.{WomExpression, PlaceholderWomExpression}
+import wdl4s.wom.graph.GraphNodePort.{GraphNodeOutputPort, OutputPort}
+import wdl4s.wom.graph._
+
+sealed trait CwlFile {
+
+  val cwlVersion: Option[CwlVersion]
+}
+
+case class Workflow(
+                     cwlVersion: Option[CwlVersion] = Option(CwlVersion.Version1),
+                     `class`: Workflow.`class`.type = Workflow.`class`,
+                     inputs: Array[InputParameter] = Array.empty,
+                     outputs: Array[WorkflowOutputParameter] = Array.empty,
+                     steps: Array[WorkflowStep]) extends CwlFile {
+
+  def womExecutable(cwlFileMap: Map[String, CwlFile]): ErrorOr[Executable] = womDefinition(cwlFileMap) map Executable.apply
+
+  def outputsTypeMap(cwlFileMap: Map[String, CwlFile]): WdlTypeMap = steps.foldLeft(Map.empty[String, WdlType]) {
+    // Not implemented as a `foldMap` because there is no semigroup instance for `WdlType`s.  `foldMap` doesn't know that
+    // we don't need a semigroup instance since the map keys should be unique and therefore map values would never need
+    // to be combined under the same key.
+    (acc, s) => acc ++ s.typedOutputs(cwlFileMap)
+  }
+
+  lazy val stepById: Map[String, WorkflowStep] = steps.map(ws => ws.id -> ws).toMap
+
+  def womGraph(cwlFileMap: Map[String, CwlFile]): ErrorOr[Graph] = {
+
+    def cwlTypeForInputParameter(input: InputParameter): Option[CwlType] = input.`type`.flatMap(_.select[CwlType])
+
+    def wdlTypeForInputParameter(input: InputParameter): Option[WdlType] = {
+      cwlTypeForInputParameter(input) map cwlTypeToWdlType
+    }
+
+    val typeMap: WdlTypeMap =
+      outputsTypeMap(cwlFileMap) ++
+        // Note this is only looking at the workflow inputs and not recursing into steps, because our current thinking
+        // is that in CWL graph inputs can only be defined at the workflow level.  It's possible that's not actually
+        // correct, but that's the assumption being made here.
+        inputs.toList.flatMap { i =>
+          wdlTypeForInputParameter(i).map(i.id -> _).toList
+        }.toMap
+
+    val graphFromInputs: Set[RequiredGraphInputNode] = inputs.map {
+      input => RequiredGraphInputNode(input.id, wdlTypeForInputParameter(input).get)
+    }.toSet
+
+    val workflowInputs: Map[String, GraphNodeOutputPort] =
+      graphFromInputs.map {
+        workflowInput =>
+          workflowInput.name -> GraphNodeOutputPort(workflowInput.name, workflowInput.womType, workflowInput)
+      }.toMap
+
+    val graphFromSteps =
+      steps.
+        toList.
+        foldLeft(Set.empty[GraphNode])(
+          (nodes, step) => step.callWithInputs(typeMap, cwlFileMap, this, nodes, workflowInputs))
+
+    val graphFromOutputs: ErrorOr[Set[GraphNode]] =
+      outputs.toList.traverse[ErrorOr, GraphNode] {
+        output =>
+
+          val tpe = cwlTypeToWdlType(output.`type`.flatMap(_.select[CwlType]).get)
```

Comment:
are you referring to `select[CwlType]`? Because those types are in fact unique to CWL and are not 1:1 to Wdl.  Thus we use `cwlTypeToWdlType` as a map to a superset.

In fact `WdlType` is an artifact of Wdl, these type should really all be `WomType` but I think there's a ticket out to do that.  If not I'll add one